File: sdk/packages/python/src/whodb/source.py
"""The whodb.source("src_...") facade: browse and read a connected source."""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from ._generated import operations as ops
from ._hydrate import hydrate_rows
from ._manifest_check import warn_if_flagged
from ._pagination import ListCall, Page

_DEFAULT_PAGE_SIZE = 100


class SourceHandle:
    """Handle for one connected data source, addressed by ID."""

    def __init__(self, execute: Callable[[ops.Request], Any], project_id: Callable[[], str], source_id: str):
        self._execute = execute
        self._project_id = project_id
        self._source_id = source_id

    def objects(self, parent: dict | None = None, page_size: int | None = None, page_offset: int | None = None) -> list[dict]:
        """List browsable objects (schemas, tables, collections...)."""
        warn_if_flagged("PlatformSourceObjects")
        return self._execute(
            ops.platform_source_objects_request(
                {
                    "projectId": self._project_id(),
                    "sourceId": self._source_id,
                    "parent": parent,
                    "kinds": None,
                    "pageSize": page_size,
                    "pageOffset": page_offset,
                }
            )
        )

    def columns(self, ref: dict) -> list[dict]:
        """List the columns of one object (table/collection)."""
        warn_if_flagged("PlatformSourceColumns")
        return self._execute(
            ops.platform_source_columns_request(
                {"projectId": self._project_id(), "sourceId": self._source_id, "ref": ref}
            )
        )

    def rows(self, ref: dict, where: dict | None = None, sort: list[dict] | None = None, page_size: int = _DEFAULT_PAGE_SIZE) -> ListCall:
        """Read rows from one object; iterate or call .pages()."""

        def fetch_page(page_offset: int) -> Page:
            warn_if_flagged("PlatformSourceRows")
            result = self._execute(
                ops.platform_source_rows_request(
                    {
                        "projectId": self._project_id(),
                        "sourceId": self._source_id,
                        "ref": ref,
                        "where": where,
                        "sort": sort,
                        "pageSize": page_size,
                        "pageOffset": page_offset,
                    }
                )
            )
            rows, total = hydrate_rows(result)
            return Page(rows, total, page_offset)

        return ListCall(fetch_page, page_size)


def list_sources(execute: Callable[[ops.Request], Any], project_id: str) -> list[dict]:
    """List the project's sources."""
    warn_if_flagged("ProjectSources")
    return execute(ops.project_sources_request({"projectId": project_id}))

File: sdk/packages/typescript/src/source.ts
import type { Transport } from './transport.js';
import * as ops from './generated/operations.js';
import type {
  Column, PlatformSource, SourceObject, SourceObjectRefInput,
  WhereCondition, SortCondition,
} from './generated/types.js';
import { hydrateRows } from './hydrate.js';
import { ListCall, type Page } from './pagination.js';
import { warnIfFlagged } from './manifest-check.js';

const DEFAULT_PAGE_SIZE = 100;

/** Options for source row reads. */
export interface SourceRowsOptions {
  where?: WhereCondition;
  sort?: SortCondition[];
  pageSize?: number;
}

/**
 * SourceHandle is the `whodb.source("src_...")` facade: browse and read a
 * connected data source by ID.
 */
export class SourceHandle {
  private readonly transport: Transport;
  private readonly projectId: () => Promise<string>;
  private readonly sourceId: string;

  constructor(transport: Transport, projectId: () => Promise<string>, sourceId: string) {
    this.transport = transport;
    this.projectId = projectId;
    this.sourceId = sourceId;
  }

  /** Lists browsable objects (schemas, tables, collections...). */
  async objects(options: { parent?: SourceObjectRefInput; pageSize?: number; pageOffset?: number } = {}): Promise<SourceObject[]> {
    warnIfFlagged('PlatformSourceObjects');
    return ops.platformSourceObjects(this.transport, {
      projectId: await this.projectId(),
      sourceId: this.sourceId,
      parent: options.parent ?? null,
      kinds: null,
      pageSize: options.pageSize ?? null,
      pageOffset: options.pageOffset ?? null,
    });
  }

  /** Lists the columns of one object (table/collection). */
  async columns(ref: SourceObjectRefInput): Promise<Column[]> {
    warnIfFlagged('PlatformSourceColumns');
    return ops.platformSourceColumns(this.transport, {
      projectId: await this.projectId(),
      sourceId: this.sourceId,
      ref,
    });
  }

  /** Reads rows from one object; supports .pages() iteration. */
  rows(ref: SourceObjectRefInput, options: SourceRowsOptions = {}): ListCall {
    const pageSize = options.pageSize ?? DEFAULT_PAGE_SIZE;
    return new ListCall(async (pageOffset): Promise<Page> => {
      warnIfFlagged('PlatformSourceRows');
      const result = await ops.platformSourceRows(this.transport, {
        projectId: await this.projectId(),
        sourceId: this.sourceId,
        ref,
        where: options.where ?? null,
        sort: options.sort ?? null,
        pageSize,
        pageOffset,
      });
      const { rows, totalCount } = hydrateRows(result);
      return { rows, totalCount, pageOffset };
    }, pageSize);
  }
}

/** Lists the project's sources. */
export async function listSources(transport: Transport, projectId: string): Promise<PlatformSource[]> {
  warnIfFlagged('ProjectSources');
  return ops.projectSources(transport, { projectId });
}
